File: LeetCode/105. Construct Binary Tree from Preorder and Inorder Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Map<Integer, Integer> inorderIndex;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        // Time and Space Complexity: O(N)

        inorderIndex = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inorderIndex.put(inorder[i], i);
        }

        return build(preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1);
    }

    private TreeNode build(int[] preorder, int[] inorder, int preStart, int preEnd, int inStart, int inEnd) {
        if (preStart > preEnd) {
            return null;
        }

        int rootVal = preorder[preStart];
        int inorderRootInd = inorderIndex.get(rootVal);

        int leftSize = inorderRootInd - inStart;
        int rightSize = inEnd - inorderRootInd;

        int preLeftEnd = preStart + leftSize;
        int preRightStart = preLeftEnd + 1;

        return new TreeNode(
            rootVal,
            build(preorder, inorder, preStart + 1, preLeftEnd, inStart, inorderRootInd - 1),
            build(preorder, inorder, preRightStart, preEnd, inorderRootInd + 1, inEnd)
        );
    }
}

File: LeetCode/105. Construct Binary Tree from Preorder and Inorder Traversal/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # Time and Space Complexity: O(N)

        inorder_index = {value: i for i, value in enumerate(inorder)}

        def build(pre_start, pre_end, in_start, in_end):
            root_val = preorder[pre_start]

            in_root_ind = inorder_index[root_val]
            left_size = in_root_ind - in_start
            right_size = in_end - in_root_ind

            left = None if left_size == 0 else build(pre_start + 1, pre_start + left_size, in_start, in_root_ind - 1)
            right = None if right_size == 0 else build(pre_start + left_size + 1, pre_end, in_root_ind + 1, in_end)

            return TreeNode(root_val, left, right)

        return build(0, len(preorder) - 1, 0, len(inorder) - 1)
